tests/form_params_test.go

```diff
@@ -0,0 +1,265 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tests
+
+import (
+	"context"
+	"go-safeweb/testing/requesttesting"
+	"log"
+	"net/http"
+	"strconv"
+	"testing"
+)
+
+func TestSimpleFormParameters(t *testing.T) {
+	reqBody := "Veg=Potato&Fruit=Apple"
+	reqBodyLen := len(reqBody)
+	postReq := []byte("POST / HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "Content-Type: application/x-www-form-urlencoded; charset=ASCII\r\n" + "Content-Length: " + strconv.Itoa(reqBodyLen) + "\r\n" + "\r\n" + reqBody + "\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		req.ParseForm()
+		if req.Form["Veg"][0] != "Potato" && req.Form["Fruit"][0] != "Apple" {
+			t.Errorf("req.Form: want Potato and Apple, got %s and %s", req.Form["Veg"][0], req.Form["Fruit"][0])
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isOkReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOK, string(resp))
+	}
+}
+
+// Test whether passing a POST request with body and without Content-Length
+// yields a 400 Bad Request
+func TestFormParametersMissingContentLength(t *testing.T) {
+	reqBody := "veggie=potato"
+	postReq := []byte("POST / HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "Content-Type: application/x-www-form-urlencoded; charset=ASCII\r\n" + reqBody + "\r\n" + "\r\n")
```

Comment:
Consider making the tests in this file table-driven as well. There seems to be a lot of common logic.

Reply:
The tests are similar but they are checking different aspects of form parameter parsing so I would rather prefer keeping them separately.